pkg/external-provider/external_series_registry.go

```diff
@@ -0,0 +1,110 @@
+package provider
+
+import (
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/provider"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/labels"
+
+	prom "github.com/directxman12/k8s-prometheus-adapter/pkg/client"
+)
+
+// ExternalSeriesRegistry acts as the top-level converter for transforming Kubernetes requests
+// for external metrics into Prometheus queries.
+type ExternalSeriesRegistry interface {
+	// ListAllMetrics lists all metrics known to this registry
+	ListAllMetrics() []provider.ExternalMetricInfo
+	QueryForMetric(namespace string, metricName string, metricSelector labels.Selector) (prom.Selector, bool, error)
+}
+
+// overridableSeriesRegistry is a basic SeriesRegistry
+type externalSeriesRegistry struct {
+	// We lock when reading/writing metrics, and rawMetrics to prevent inconsistencies.
+	mu sync.RWMutex
+
+	// metrics is the list of all known metrics, ready to return from the API
+	metrics []provider.ExternalMetricInfo
+	// rawMetrics is a lookup from a metric to SeriesConverter for the sake of generating queries
+	rawMetrics map[string]SeriesConverter
+
+	mapper apimeta.RESTMapper
+
+	metricLister MetricListerWithNotification
```

Comment:
@john-delivuk I might have missed something, but is this used somewhere?

Reply:
I need to dig in, but I believe this is tied to how updates to the Metrics name cache are handled.

Reply:
As commented in https://github.com/DirectXMan12/k8s-prometheus-adapter/pull/146#discussion_r259388559, the constructor of the series registry just installs a notification receiver callback to be invoked when metrics are being listed periodically.

There is no need to have it here as a struct member.